flink-core/src/test/java/org/apache/flink/api/dag/TransformationTest.java

```diff
@@ -45,7 +47,25 @@ public class TransformationTest extends TestLogger {
 
     @Before
     public void setUp() {
-        transformation = new TestTransformation<>("t", null, 1);
+        transformation = new TestTransformation<>("t", null, 1, null);
+    }
+
+    @Test
+    public void testPredecessorCache() throws Exception {
+        // diamond-like transformation:
+        //                   --> midNode-\
+        //              ---/             --\
+        //           --/                    ->
+        //      commonNode  -----------------> topNode
+        Transformation<Void> commonNode = new TestTransformation<>("commonNode", null, 1, null);
+        Transformation<Void> midNode =
+                new TestTransformation<>("midNode", null, 1, Collections.singletonList(commonNode));
+        Transformation<Void> topNode =
+                new TestTransformation<>("topNode", null, 1, Arrays.asList(midNode, commonNode));
+
+        topNode.getTransitivePredecessors();
+        assertEquals(topNode.getPredecessorsCache().size(), 1);
```

Comment:
Flink is turning to JUnit5. According to the community guide[1], new tests should be written using JUnit5. If we have to modify an existing testing class, we should migrate it to JUnit5 first in a hotfix commit, for example, [#20350](https://github.com/apache/flink/pull/20350/commits).

[1] https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#tooling